tools/internal/chartfetch/cmd_test.go

```diff
@@ -0,0 +1,100 @@
+package chartfetch
+
+import (
+	"errors"
+	"fmt"
+	"github.com/broadinstitute/terra-helmfile-images/tools/internal/shellmock"
+	"github.com/stretchr/testify/assert"
+	"regexp"
+	"strings"
+	"testing"
+)
+
+// Integration test for `chartfetch`
+// Executes the Cobra command with the given arguments and verifies that the
+// correct "helm fetch" command is executed under the hood.
+func TestFetch(t *testing.T) {
+	tmpDir := t.TempDir()
+
+	testCases := []struct {
+		description   string
+		args          []string
+		setupMocks    func(t *testing.T, m *shellmock.MockRunner)
+		expectedError *regexp.Regexp
+	}{
+		{
+			description:   "no arguments",
+			expectedError: regexp.MustCompile(`accepts 1 arg\(s\), received 0`),
+		},
+		{
+			description:   "missing -d flag",
+			args:          args("terra-helm/leonardo -v 1.2.3"),
+			expectedError: regexp.MustCompile(`required flag\(s\) "download-dir" not set`),
+		},
+		{
+			description:   "missing -v flag",
+			args:          args("terra-helm/leonardo -d /does/not/exist"),
+			expectedError: regexp.MustCompile(`required flag\(s\) "version" not set`),
+		},
+		{
+			description: "should exit without downloading if directory exists",
+			args:        args("terra-helm/leonardo -v 1.2.3 -d %s", tmpDir),
+		},
+		{
+			description: "should download if directory does not exist",
+			args:        args("terra-helm/leonardo -v 1.2.3 -d %s/download-dir", tmpDir),
+			setupMocks: func(t *testing.T, m *shellmock.MockRunner) {
+				m.ExpectCmdFmt(t, "helm fetch terra-helm/leonardo --untar -d %s/download-dir", tmpDir)
+			},
+		},
+		{
+			description: "should return an error if the helm fetch command fails",
+			args:        args("terra-helm/leonardo -v 1.2.3 -d %s/download-dir", tmpDir),
+			setupMocks: func(t *testing.T, m *shellmock.MockRunner) {
+				m.ExpectCmdFmt(t, "helm fetch terra-helm/leonardo --untar -d %s/download-dir", tmpDir).
+					Return(errors.New("command failed because reasons"))
+			},
+			expectedError: regexp.MustCompile("command failed because reasons"),
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.description, func(t *testing.T) {
+			// Create a new mock runner and inject it by setting shellRunner package-level variable to mock
+			var mockRunner = shellmock.DefaultMockRunner()
+			var originalRunner = shellRunner
+			shellRunner = mockRunner
+			defer func() { shellRunner = originalRunner }()
```

Comment:
I'm not sure I understand exactly what this part does or why it needs to be deferred?

Reply:
This is an area where suggestions would be appreciated, actually.

I prefer end-to-end testing for CLI tools. I.e., I construct a new Cobra command, set some CLI args, and execute it, and then verify that the expected effects (outgoing HTTP requests, shell commands, etc) have occurred. This way a simple API/signature change to an internal method won't require me to update a bunch of unit tests.

The downside to this approach is that injecting mocks is difficult, because the Cobra command is responsible for constructing objects and wiring them together. The approach I've used here is to store a default shellRunner in a package-level variable and then shadow it with a mock during tests.

That's what's happening there -- `shellRunner` is a package-level variable, not local. The package-level variable is being replaced with a mock and restored once the tests are finished.

I hate "global" state as much as the next person. It certainly limits our ability to run tests for this package in parallel. But I also wanted to keep Cobra command initialization simple and this was the best option I could come up with! And we'd run into the same issue using https://github.com/jarcoal/httpmock. :shrug: